cranelift/filetests/src/function_runner.rs

```diff
@@ -1,197 +1,344 @@
+            self.define_function(func.clone())?;
+            self.create_trampoline_for_function(func)?;
+        }
+
+        Ok(())
+    }
+
+    /// Declares a function an registers it as a linkable and callable target internally
+    pub fn declare_function(&mut self, func: &Function) -> Result<()> {
+        let next_id = self.defined_functions.len() as u32;
+        match self.defined_functions.entry(func.name.clone()) {
+            Entry::Occupied(_) => {
+                anyhow::bail!("Duplicate function with name {} found!", &func.name)
+            }
+            Entry::Vacant(v) => {
+                let name = func.name.to_string();
+                let func_id =
+                    self.module
+                        .declare_function(&name, Linkage::Local, &func.signature)?;
+
+                v.insert(DefinedFunction {
+                    new_name: UserExternalName::new(TESTFILE_NAMESPACE, next_id),
+                    signature: func.signature.clone(),
+                    func_id,
+                });
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Renames the function to its new [UserExternalName], as well as any other function that
+    /// it may reference.
+    ///
+    /// We have to do this since the JIT cannot link Testcase functions.
+    fn apply_func_rename(
+        &self,
+        mut func: Function,
+        defined_func: &DefinedFunction,
+    ) -> Result<Function> {
+        // First, rename the function
+        let func_original_name = func.name;
+        func.name = UserFuncName::User(defined_func.new_name.clone());
+
+        // Rename any functions that it references
+        // Do this in stages to appease the borrow checker
+        let mut redefines = Vec::with_capacity(func.dfg.ext_funcs.len());
+        for (ext_ref, ext_func) in &func.dfg.ext_funcs {
+            let old_name = match &ext_func.name {
+                ExternalName::TestCase(tc) => UserFuncName::Testcase(tc.clone()),
+                ExternalName::User(username) => {
+                    UserFuncName::User(func.params.user_named_funcs()[*username].clone())
+                }
+                // The other cases don't need renaming, so lets just continue...
+                _ => continue,
+            };
+
+            let target_df = self.defined_functions.get(&old_name).ok_or(anyhow!(
+                "Undeclared function {} is referenced by {}!",
+                &old_name,
+                &func_original_name
+            ))?;
+
+            redefines.push((ext_ref, target_df.new_name.clone()));
+        }
+
+        // Now register the redefines
+        for (ext_ref, new_name) in redefines.into_iter() {
+            // Register the new name in the func, so that we can get a reference to it.
+            let new_name_ref = func.params.ensure_user_func_name(new_name);
+
+            // Finally rename the ExtFunc
+            func.dfg.ext_funcs[ext_ref].name = ExternalName::User(new_name_ref)
```

Comment:
semicolon at end of statement? (I guess here rustc let it slide because it's the end of the block)